### PR TITLE
fix(levm): fix interaction with cache and db

### PR DESCRIPTION
**Pull request by JereSalo**
**Motivation**

**Description**

TLDR: We don't want to interact directly with the db without consulting before with the cache. This won't fix nor break tests because in every EFTest we send the cache empty, but it is a necessary fix for executing multiple transactions within a block. As a side effect we have to move contract creation and insertion to cache to `transact()`, so that we can revert if address is already occupied.

Idea:
- Move contract creation and insertion to cache to `transact()`, not to `new()`.
    - This is because we really want to check if the contract already exists both in the cache and in the db, and if when you want to create it you see that it already exists in one of those 2 there you revert. Currently we only check it in the db but the tests work because they test only one transaction each, so the cache always starts empty.
- The idea is to stop using `db.get_account_info()` in the `VM::new()`, we want to query the cache first and then the DB. For this we have the `get_account()` method in VM, but in `VM::new()` it doesn't work because there is no vm created yet. The idea is to grab that behavior and take it to a `get_account(address, mut cache, db)` function that has the same behavior as now. The VM get_account method can be a handrail to call this, which does `get_account(address, self.cache, self.db)`.

Additional Changes:
- Remove unnecessary stuff in `VM::new()`
- Remove receiver account every time a transaction reverts, independently of it's type.

Closes #issue_number

### CODE REVIEW

File: crates/vm/levm/src/vm.rs [hunk 1]
```diff
@@ -1208,3 +1172,19 @@ fn get_number_of_topics(op: Opcode) -> Result<u8, VMError> {
...
+
+/// Gets account, first checking the cache and then the database (caching in the second case)
+pub fn get_account(cache: &mut CacheDB, db: &Arc<dyn Database>, address: Address) -> Account {
+    match cache::get_account(cache, &address) {
+        Some(acc) => acc.clone(),
+        None => {
+            let account_info = db.get_account_info(address);
+            let account = Account {
+                info: account_info,
+                storage: HashMap::new(),
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/vm.rs:1184]: Hi Jere, quick question, why are we always creating a new storage every time we "get" an account?

Comment (JereSalo) [reply crates/vm/levm/src/vm.rs:1184]: Because we actually want to get only the account info but manipulating the account itself is sometimes more comfortable. When we get the account info we want to insert that into cache, but it stores Accounts, not account info. So what we do is inserting the `Account` with the corresponding `AccountInfo` but with empty storage, and if we need the storage slot we have to access the Database.

Comment (lima-limon-inc) [reply crates/vm/levm/src/vm.rs:1184]: >  (...) and if we need the storage slot we have to access the Database.

Do you know where this is done? I am not quite clear on why we are doing it in two steps.

**Review by lima-limon-inc — APPROVED** [review]
LGTM

**Review by damiramirez — APPROVED** [review]
LGTM

File: crates/vm/levm/src/vm.rs [hunk 2]
```diff
@@ -1156,18 +1131,7 @@ impl VM {
...
-                let account_info = self.db.get_account_info(address);
-                let account = Account {
-                    info: account_info,
-                    storage: HashMap::new(),
-                };
-                cache::insert_account(&mut self.cache, address, account.clone());
-                account
-            }
-        }
+        get_account(&mut self.cache, &self.db, address)
```
Comment (maximopalopoli) [inline crates/vm/levm/src/vm.rs:1134]: Maybe we could have a different name for the two functions, since having the same one could lead to confusion.

**Review by maximopalopoli — APPROVED** [review]
LGTM

File: crates/vm/levm/src/vm.rs (hunk 2 above)
Comment (JereSalo) [reply crates/vm/levm/src/vm.rs:1134]: hmm maybe, the parameters are different though. They have the same name because they do the same thing but the `VM` one is like a wrapper. A better name doesn't come to my mind, what do you suggest?

Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs:1134]: It's true that there isn't a good name for both, maybe it could be get_account_cache_or_db, but it says more about implementation than anything else. Actual version looks good to me in that sense 👍

**Review by ilitteri — APPROVED** [review]
